File: MIT/neural_network.py
import numpy as np


class Layer:
    def __init__(self, n_inputs, n_neurons):
        self.weights = 0.10 * np.random.randn(n_inputs, n_neurons) # initialize a matrix n_inputs x n_neurons
        self.biases = np.zeros((1, n_neurons))

    def call(self, inputs):
        z = np.matmul(inputs, self.weights) + self.biases
        output = sigmoid(z)
        return output


def sigmoid(z):
    return 1 / (1 + np.exp(-z))

File: CPP_NN/neural_network.cpp
#include <iostream>
#include <vector>
#include <cmath>

using namespace std;

double sigmoid(double x) {
    return 1 / (1 + exp(-x));
}


class Layer {
    public:
        vector<vector<double>> weights;
        vector<double> biases;
        vector<double> outputs;
        
        Layer(int n_inputs, int n_neurons) {
            for (int i = 0; i < n_neurons; i++) {
                vector<double> neuron_weights;
                for (int j = 0; j < n_inputs; j++) {
                    neuron_weights.push_back(0.0);
                }
                weights.push_back(neuron_weights);
                biases.push_back(0.0);
                outputs.push_back(0.0);
            }
        }

        void call(vector<vector<double>> inputs) {
            for (int i = 0; i < weights.size(); i++) {
                double output = 0.0;
                for (int j = 0; j < weights[i].size(); j++) {
                    output += inputs[0][j] * weights[i][j];
                }
                output += biases[i];
                // apply sigmoid function
                output = sigmoid(output);
                outputs[i] = output;
            }
        }
};
